### ISSUE TITLE
Main Assignment: ML Project Implementation in AWS Ecosystem

### ISSUE THREAD

**Issue by sprince0031** [sprince0031/UL_Masters#105]
Find project details in the [Brightspace assignment submission page](https://learn.ul.ie/d2l/lms/dropbox/user/folder_submit_files.d2l?ou=17937&isprv=0&db=22988&cfql=1).  
  
## PDF report
PDF report must be based on [assignment template](https://learn.ul.ie/d2l/common/dialogs/quickLink/quickLink.d2l?ou=17937&type=coursefile&fileId=Main+Assignment+Template%2fCS5024+Assignment+Template.docx).

### To do:
- [x] Submission file renamed to format \<Student ID number>\_\<Surname>\_\<Firstname>.pdf
- [x] Edit out the advisory red text provided by the template after using it to build the report.
- [x] An abstract outlining the problem you are addressing.
- [x] An introduction which elaborates on the problem.
- [x] A description of the AI Ecosystem that you have created.
- [x] A description and elaboration of the model.
- [x] Examples of results obtained from when you ran your model in the AI Ecosystem, along with illustrative application (e.g., accuracy, precision, recall) and cloud KPIs, describing any challenges you had in the implementation.
- [x] An elaboration on how you would modify the ecosystem should hypothetical expansions of the application's scope occur. You should reference the AWS Advisory tools and ecosystem pillars to do so, as described in the course materials provided through Brightspace.
- [x] Demonstrate within your submission report that your code runs in the environment you have presented.

## Code zip file
- [x] Submission file renamed to format \<Student ID number>\_\<Surname>\_\<Firstname>\_code.zip

## Forum engagement (10%)
- [x] Actively post and discuss on the main assignment forum.

> [!important]
> Due by 2<sup>nd</sup> May, 2024 23:59

**Comment by sprince0031** [comment]
## Project steps
- [ ] Set up an ML SageMaker model.
  - [ ] Store data
  - [ ] train model
  - [ ] Deploy model

- [ ] Setup simple web app
  - [ ] Deploy web page code.
  - [ ] Have a form input with image upload
    - [ ] API POST requests sends image to Lambda
      - [ ] Lambda stores image to DB with "retrained" Boolean field set to False.
      - [ ] Response is the classification prediction.
      - [ ] Have another input to verify the prediction
      - [ ] API POST request #2 sends back the verification (human label)
        - [ ] Updates DB record with the predicted and human classification labels. 

- [ ] Create lambda function to take image input
  - [ ] Send it to deployed pickle model
  - [ ] Get classification output
- [ ] Create lambda function to take the human verification of classification input
  - [ ] Update the db and check for the retrain batch threshold.
    - [ ] If the number of new records is equal to the minimum retrain batch size, initiate retraining of the deployed ML model with the new training data. 
    - [ ] Also check if the accuracy of the new data has dropped below the required threshold. Retraining is required only if this drops.

**Comment by sprince0031** [comment]
> [!warning]
> Submitted 24 hours late on 04/05/2024 00:00.
